src/mds/MDSCacheObject.cc

```diff
@@ -85,52 +85,30 @@ void MDSCacheObject::dump_states(ceph::Formatter *f) const
     f->dump_string("state", "rejoinundef");
 }
 
-bool MDSCacheObject::is_waiter_for(uint64_t mask, uint64_t min) {
-  if (!min) {
-    min = mask;
-    while (min & (min-1))  // if more than one bit is set
-      min &= min-1;        //  clear LSB
-  }
-  for (auto p = waiting.lower_bound(min); p != waiting.end(); ++p) {
-    if (p->first & mask) return true;
-    if (p->first > mask) return false;
+bool MDSCacheObject::is_waiter_for(waitmask_t mask) {
+  for ([[maybe_unused]] auto& [seq, waiter] : waiting) {
+    if ((waiter.mask & mask).any()) {
+      return true;
+    }
   }
   return false;
 }
 
-void MDSCacheObject::take_waiting(uint64_t mask, MDSContext::vec& ls) {
-  if (waiting.empty()) return;
-
-  // process ordered waiters in the same order that they were added.
-  std::map<uint64_t, MDSContext*> ordered_waiters;
-
+void MDSCacheObject::take_waiting(waitmask_t mask, MDSContext::vec& ls) {
```

Comment:
Unfortunately, this is not backward compatible wrt non-ordered events.

Previously, non-ordered waiters would be called in the bitmask priority order before all ordered waiters.
After this refactoring the non-ordered waiters will be processed close to the insertion order between two ordered waiters, making the bitmask priority not functional.

Reply:
Hmm? The seq for waiters without `WAIT_ORDERED` is `0` so they are always at the front (of iteration) of the `std::compact_multimap`?

Reply:
You are right! 
I misread the changed function in the diff, and thought you were adding with the latest seq. Sorry, all is good now